packages/animations/lib/src/utils/composite_animation_widget.dart

```diff
@@ -0,0 +1,253 @@
+  /// has visible state same as the animation widgets defined in
+  /// [enterTransitionBuilder].
+  ///
+  /// Defined true at [FadeScaleTransition] since it only used one animation.
+  final bool syncVisibleStates;
+
+  @override
+  _CompositeAnimationWidgetState createState() =>
+      _CompositeAnimationWidgetState();
+}
+
+class _CompositeAnimationWidgetState extends State<CompositeAnimationWidget> {
+  AnimationStatus _effectiveAnimationStatus;
+  Animation<double> _enterAnimation;
+  Animation<double> _exitAnimation;
+
+  @override
+  void initState() {
+    super.initState();
+    widget.animation.addStatusListener(_animationListener);
+    _effectiveAnimationStatus = widget.animation.status;
+    _setStartingAnimationValues(widget.flip);
+  }
+
+  void _animationListener(AnimationStatus animationStatus) {
+    _effectiveAnimationStatus = _calculateEffectiveAnimationStatus(
+      lastEffective: _effectiveAnimationStatus,
+      current: animationStatus,
+    );
+    // change the animation values based on effectiveAnimationStatus.
+    _setStartingAnimationValues(widget.flip);
+  }
+
+  // This is called at [initState], [didUpdateWidget], and [_animationListener]
+  void _setStartingAnimationValues(bool flip) {
+    final Animation<double> baseAnimation =
+        flip ? _flip(widget.animation) : widget.animation;
+    _enterAnimation = flip ? _flip(baseAnimation) : baseAnimation;
+    if (widget.syncVisibleStates) {
+      _exitAnimation = _enterAnimation;
+    } else {
+      _exitAnimation = flip ? baseAnimation : _flip(baseAnimation);
+    }
+    assert(_effectiveAnimationStatus != null);
+    assert(_enterAnimation != null);
+    assert(_exitAnimation != null);
+    switch (_effectiveAnimationStatus) {
+      case AnimationStatus.forward:
+        // when playing forward, the exitTransition must be visible.
+        if (widget.syncVisibleStates) {
+          _exitAnimation = flip
+              ? const AlwaysStoppedAnimation<double>(0.0)
+              : const AlwaysStoppedAnimation<double>(1.0);
+        } else {
+          _exitAnimation = flip
+              ? const AlwaysStoppedAnimation<double>(1.0)
+              : const AlwaysStoppedAnimation<double>(0.0);
+        }
+        break;
+      case AnimationStatus.dismissed:
+      case AnimationStatus.reverse:
+      case AnimationStatus.completed:
+        // when playing reverse, the enterTransition must be visible.
+        _enterAnimation = flip
+            ? const AlwaysStoppedAnimation<double>(0.0)
+            : const AlwaysStoppedAnimation<double>(1.0);
+    }
+  }
+
+  @override
+  void didUpdateWidget(CompositeAnimationWidget oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.animation != widget.animation) {
+      oldWidget.animation.removeStatusListener(_animationListener);
+      widget.animation.addStatusListener(_animationListener);
+      _animationListener(widget.animation.status);
+    }
+    if (widget.flip != oldWidget.flip) {
```

Comment:
what if syncVisibleStates changes?

Reply:
The default flow of values in those animations objects if `syncVisibleStates` is false
_enterAnimation: 0.0  -> 1.0
_exitAnimation: 1.0 -> 0.0

in `syncVisibleState` is true

_enterAnimation: 0.0  -> 1.0
_exitAnimation: 0.0 -> 1.0 (syncs to enterAnimation, same visibleState)

just to clarify the _exitAnimation is always flip against _enterAnimation and all "End" animation values are always the visible states, "flip" flag will flip all of the values mentioned above.

the only usage of `syncVisibleStates` set to true is in `FadeScaleTransition` where in the original implementation the exit animation is not flipped.